### PR TITLE
Implements equals method on AbstractChange to make sure that we can compare Changes correctly

### PR DESCRIPTION
**Pull request by filipelautert**
## Impact

- [X] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description

Commit https://github.com/liquibase/liquibase/pull/3616/commits/279b1b35d9c5f984d2f8542b792c870fd71c3ec0# that is part of checksums upgrade moved the dbms logic from Changeset to DbmsFilter. But it happens that Change object and all of the inheriting classes to not implement equals method. So when it calls collection.removeAll(object) it is ending up removing all changes that are of the specific type (for instance AbstractSqlChange), and not only the expected change. 

So if you execute a sample changelog like the one below in a `H2 `database:

```
    <changeSet id="0" author="fl" >
        <sql>create table anydb(id boolean default true, status varchar(2))</sql>
        <sql dbms="oracle">create table oraculo(id boolean default true, status varchar(2))</sql>
        <dropTable tableName="anydb" />
    </changeSet>
```

The expected result would be to create anydb table then drop it; but it is just dropping the table.

This PR fixes this issue.

### CODE REVIEW

**Review by abrackx — COMMENTED** [review]
Can we add that example as a test case?

File: liquibase-standard/src/main/java/liquibase/change/AbstractChange.java [hunk 1]
```diff
@@ -868,4 +868,21 @@ public String getDescription() {
...
     }
 
+    @Override
+    public boolean equals(Object reference) {
+        if (this == reference) return true;
+        if (reference instanceof DbmsTargetedChange && this instanceof DbmsTargetedChange &&
+            !((DbmsTargetedChange)reference).getDbms().equals(((DbmsTargetedChange)this).getDbms())) {
+            return false;
+        }
+        if (reference == null || getClass() != reference.getClass()) return false;
```
Comment (wwillard7800) [inline liquibase-standard/src/main/java/liquibase/change/AbstractChange.java:878]: Does this null check need to happen right after the same-object check to avoid an NPE?  The instanceof call will allow it a null.

I know that is an edge case.

**Review by wwillard7800 — APPROVED** [review]

**Comment by MalloD12** [comment]
Hey @rberezen I have added a test to ensure a case like the one Filipe mentioned above doesn't happen again. We only limited this scope to H2, because we just want to test a dbms filtered change that is not deployed in any other given DB (H2 in this case). Are you ok with this or think this should be more generic and be added for example in **AbstractIntegrationTest** or somewhere else?

File: liquibase-standard/src/main/java/liquibase/change/AbstractChange.java (hunk 1 above)
Comment (filipelautert) [reply liquibase-standard/src/main/java/liquibase/change/AbstractChange.java:878]: I simulated it here, but "reference instanceof DbmsTargetedChange" returns false so the second if will bail out on the first condition and won't call the getDbms methods.

**Review by rberezen — CHANGES_REQUESTED** [review]
hi @filipelautert 
fails for me when 

```
    <changeSet id="0" author="fl" >
        <sql dbms="oracle">create table anydb(id boolean default true, status varchar(2))</sql>
        <sql dbms="oracle">create table oraculo(id boolean default true, status varchar(2))</sql>
        <dropTable tableName="anydb" />
    </changeSet>
```

```
Unexpected error running Liquibase: java.lang.StackOverflowError

java.lang.StackOverflowError
	at java.base/java.util.TreeMap$KeyIterator.<init>(TreeMap.java:1259)
	at java.base/java.util.TreeMap.keyIterator(TreeMap.java:1107)
	at java.base/java.util.TreeMap$KeySet.iterator(TreeMap.java:1120)
	at java.base/java.util.AbstractCollection.addAll(AbstractCollection.java:351)
	at java.base/java.util.TreeSet.addAll(TreeSet.java:312)
	at java.base/java.util.TreeSet.<init>(TreeSet.java:160)
	at liquibase.AbstractExtensibleObject.getAttributes(AbstractExtensibleObject.java:41)
	at liquibase.util.StringUtil.join(StringUtil.java:390)
	at liquibase.AbstractExtensibleObject.describe(AbstractExtensibleObject.java:238)
	at liquibase.AbstractExtensibleObject.equals(AbstractExtensibleObject.java:259)
	at liquibase.change.AbstractChange.equals(AbstractChange.java:879)
	at java.base/java.util.Objects.equals(Objects.java:77)
```

**Comment by rberezen** [comment]
> Hey @rberezen I have added a test to ensure a case like the one Filipe mentioned above doesn't happen again. We only limited this scope to H2, because we just want to test a dbms filtered change that is not deployed in any other given DB (H2 in this case). Are you ok with this or think this should be more generic and be added for example in **AbstractIntegrationTest** or somewhere else?

I'm fine with that. Could you pleas add a scenario for this changeset?  Thanks in advance
```
    <changeSet id="0" author="fl" >
        <sql dbms="oracle">create table anydb(id boolean default true, status varchar(2))</sql>
        <sql dbms="oracle">create table oraculo(id boolean default true, status varchar(2))</sql>
        <dropTable tableName="anydb" />
    </changeSet>
```

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java [hunk 2]
```diff
@@ -189,6 +190,22 @@
         Assert.assertEquals(insertedValue.toString(), queryResult.get(1).get(colName));
     }
 
+    @Test
+    public void makeSureDbmsFilteredChangeIsNotDeployed() throws Exception {
+        clearDatabase();
+
+        Liquibase liquibase = createLiquibase("changelogs/h2/complete/sql.change.dbms.filtered.should.not.be.deployed.changelog.xml");
+        liquibase.update();
```
[the remark this answers is not in the document]
Comment (rberezen) [reply liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java:198]: @MalloD12 could you take a looks on this please?

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java [hunk 3]
```diff
@@ -189,6 +190,22 @@ public void testRollbackWithoutContext() throws Exception {
...
```
Comment (rberezen) [inline liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java:198]: @MalloD12 if `liquibase.update();` fails, test will also fail, right?

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (MalloD12) [reply liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java]: Added a new `runUpdate()` method in the AbstractIntegrationTest class to run the update command without having to use the deprecated one anymore.

**Comment by MalloD12** [comment]
> > Hey @rberezen I have added a test to ensure a case like the one Filipe mentioned above doesn't happen again. We only limited this scope to H2, because we just want to test a dbms filtered change that is not deployed in any other given DB (H2 in this case). Are you ok with this or think this should be more generic and be added for example in **AbstractIntegrationTest** or somewhere else?
> 
> I'm fine with that. Could you pleas add a scenario for this changeset? Thanks in advance
> 
> ```
>     <changeSet id="0" author="fl" >
>         <sql dbms="oracle">create table anydb(id boolean default true, status varchar(2))</sql>
>         <sql dbms="oracle">create table oraculo(id boolean default true, status varchar(2))</sql>
>         <dropTable tableName="anydb" />
>     </changeSet>
> ```

Added this new changeset, but I removed the `dropTable` change of it, in this case, it shouldn't be there unless we want to make sure the test fails trying to remove `anydb` table. If that's the case, I'm already covering that part with a `try-catch` by executing a `select` against that table. 

Please have a look at it and let me know if you are ok with the changes made.

Thanks,
Daniel.

File: liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java (hunk 3 above)
Comment (MalloD12) [reply liquibase-integration-tests/src/test/java/liquibase/dbtest/h2/H2IntegrationTest.java]: I think it would depend on the failure it returns, if there is a validation error but it doesn't stop the execution of the update then it will try to run the rest of the tests, but if for some reason `update` execution is stopped by an exception test will fail as well.

**Review by rberezen — APPROVED** [review]
LGTM! Thanks @MalloD12 and @filipelautert!
